projects/step-core/src/lib/client/augmented/services/augmented-resources-service.ts

```diff
@@ -14,7 +16,96 @@ export class AugmentedResourcesService extends ResourcesService {
     id: 'id',
   });
 
-  constructor(httpRequest: BaseHttpRequest, private _tableRest: TableApiWrapperService) {
-    super(httpRequest);
+  constructor(
+    public _httpRequest: BaseHttpRequest,
+    public _tableRest: TableApiWrapperService,
+    public _httpClient: HttpClient
+  ) {
+    super(_httpRequest);
+  }
+
+  /**
+   * Upload a resource while keeping track of the progress
+   * @note $progress and $response are hot observables, we have to make sure to unsubscribe from them
+   */
+  createResourceWithProgress({
+    file,
+    queryParams: { type, duplicateCheck, directory },
+    resourceId,
+  }: {
+    file: File;
+    queryParams: {
+      type: string;
+      duplicateCheck: boolean;
+      directory: boolean;
+    };
+    resourceId?: string;
+  }): {
+    progress$: Observable<number>;
+    response$: Observable<ResourceUploadResponse>;
+  } {
+    const formData = new FormData();
+
+    formData.set('file', file);
+
+    const stream$ = this._httpClient
+      .request('POST', `/rest/resources${resourceId ? `/${resourceId}` : ''}/content`, {
+        body: formData,
+        params: {
+          type,
+          duplicateCheck,
+          directory,
+        },
+        headers: {
+          enctype: 'multipart/form-data',
+        },
+        observe: 'events',
+        responseType: 'arraybuffer',
+        reportProgress: true,
+      })
+      .pipe(shareReplay());
+
+    const progress$ = stream$.pipe(
+      filter((httpEvent) => httpEvent.type === HttpEventType.UploadProgress),
+      switchMap((httpEvent) => {
+        const httpProgressEvent = httpEvent as HttpProgressEvent;
+
+        const progressPercentage = this.calculateProgressPercentage(httpProgressEvent);
+
+        return of(progressPercentage);
+      })
+    );
+
+    const response$ = stream$.pipe(
+      filter((httpEvent) => httpEvent.type === HttpEventType.Response),
+      switchMap((httpEvent) => {
+        const httpResponse = httpEvent as HttpResponse<ArrayBuffer>;
+
+        if (!httpResponse.body) {
+          return of();
+        }
+
+        const resourceUploadResponse = this.decodeArrayBuffer<ResourceUploadResponse>(httpResponse.body);
+
+        return of(resourceUploadResponse);
+      })
+    );
+
+    return {
+      progress$,
+      response$,
+    };
+  }
+
+  downloadResource(resourceId: string): void {
+    window.location.href = `/rest/resources/${resourceId}/content`;
```

Comment:
API service shouldn't change the location.
It should just return the url.

I would prefer to keep the anchor, so we could additionally setup the target, or do any further manipulations with url